File: src/Core/providers.ts
import { SectionConfig } from './models/SectionConfig';
import { ISecretProvider } from './providers/ISecretProvider';
import axios from "axios";

export class Providers {
  private secretProvider;

  constructor(secretProvider: ISecretProvider) {
    this.secretProvider = secretProvider;
  }

  async sendRequestToModel(prompt: string, content: string, config: SectionConfig) {

    if (config.values.useOllama) {
      return this.useLocalOpenAI(prompt, content, config);
    }
    else {
      return this.useAzureOpenAI(prompt, content, config);
    }
  }

  /**
   * Use Local OpenAI Compliant API to generate summary
   *
   * @param {string} prompt
   * @param {string} content
   * @param {string} fileName
   * @return {*} 
   */
  private async useLocalOpenAI(prompt: string, content: string, config: SectionConfig) {
    let settings: any = {
      "endpoint": config.values.modelEndpoint,
      "apikey": await this.secretProvider.getSecret('model-local-api-endpoint-key'),
      "modelName": config.values.modelName,
    };
    var options = {
      method: 'POST',
      url: `${settings.endpoint}/api/generate`,
      headers: { 'api-key': settings.apikey },
      data: {
        model: settings.modelName,
        num_gpu: 0,
        main_gpu: 0,
        stream: false,
        prompt: content,
        system: prompt
      }
    };

    try {
      var response = await axios.request(options);
      console.log(response.data.response);
      return response.data.response;
    }
    catch (error) {
      console.log(error);
      return error;
    }
  }

  private async useAzureOpenAI(prompt: string, content: string, config: SectionConfig) {
    let settings = {
      "endpoint": config.values.modelEndpoint,
      "apiversion": config.values.modelVersion,
      "model": config.values.modelName,
      "apikey": await this.secretProvider.getSecret('model-azure-openai-model-api-endpoint-key'),
    };
    var options = {
      method: 'POST',
      url: `${settings.endpoint}/openai/deployments/${settings.model}/chat/completions`,
      params: { 'api-version': settings.apiversion },
      headers: { 'api-key': settings.apikey },
      data: {
        messages: [
          {
            role: 'system',
            content: [
              {
                type: 'text',
                text: prompt
              }
            ]
          },
          { role: 'user', content: [{ type: 'text', text: content }] }
        ]
      }
    };

    try {
      var response = await axios.request(options);
      return response.data.choices[0].message.content;
    }
    catch (error) {
      console.log(error);
      return error;
    }
  }
}

File: src/Extensions/VisualStudio/src/DocuGen/Providers.cs
using Newtonsoft.Json;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;

public class Providers
{
    private readonly ISecretProvider _secretProvider;
    private readonly HttpClient _httpClient;

    public Providers(ISecretProvider secretProvider)
    {
        _secretProvider = secretProvider;
        _httpClient = new HttpClient();
    }

    public async Task<string> SendRequestToModelAsync(string prompt, string content, ModelProviderEnums modelProvider, string modelEndpoint, string modelName, string modelVersion)
    {
        switch (modelProvider)
        {
            case ModelProviderEnums.Ollama:
                return await UseOllamaAsync(prompt, content, modelEndpoint, modelName, modelVersion);
            case ModelProviderEnums.AzureOpenAI:
                return await UseAzureOpenAIAsync(prompt, content, modelEndpoint, modelName, modelVersion);
            default:
                throw new Exception($"Unsupported model provider: {modelProvider}");
        }
    }

    private async Task<string> UseOllamaAsync(string prompt, string content, string modelEndpoint, string modelName, string modelVersion)
    {
        var settings = new
        {
            endpoint = modelEndpoint,
            apikey = await _secretProvider.GetSecret("modelApiKey"),
            modelName = modelName
        };

        var requestBody = new
        {
            model = settings.modelName,
            num_gpu = 0,
            main_gpu = 0,
            stream = false,
            prompt = content,
            system = prompt
        };

        var response = await _httpClient.PostAsync($"{settings.endpoint}/api/generate", new StringContent(JsonConvert.SerializeObject(requestBody), Encoding.UTF8, "application/json"));
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    private async Task<string> UseAzureOpenAIAsync(string prompt, string content, string modelEndpoint, string modelName, string modelVersion)
    {
        var settings = new
        {
            endpoint = modelEndpoint,
            apiversion = modelVersion,
            model = modelName,
            apikey = await _secretProvider.GetSecret("modelApiKey")
        };

        var requestBody = new
        {
            messages = new[]
            {
                new
                {
                    role = "system",
                    content = new[]
                    {
                        new { type = "text", text = prompt }
                    }
                },
                new
                {
                    role = "user",
                    content = new[]
                    {
                        new { type = "text", text = content }
                    }
                }
            }
        };

        _httpClient.DefaultRequestHeaders.Add("api-key", settings.apikey);

        var response = await _httpClient.PostAsync(
            $"{settings.endpoint}/openai/deployments/{settings.model}/chat/completions?api-version={settings.apiversion}",
            new StringContent(JsonConvert.SerializeObject(requestBody), Encoding.UTF8, "application/json")
        );
        
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }
}